Numpy is the base library for many other libraries such as Pandas, OpenCV. It deals with the numbers as arrays of numbers, images in which each pixel is made up of arrays of numbers, etc.

In [2]:
import numpy

n = numpy.arange(27)     #converts to a one dimensional array
n

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26])

In [4]:
type(n)     # n is a n dimensional array

numpy.ndarray

In [5]:
# reshape is used to convert arrays in different dimensions
n.reshape(3, 9)    # converting to two dimensional array


array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8],
       [ 9, 10, 11, 12, 13, 14, 15, 16, 17],
       [18, 19, 20, 21, 22, 23, 24, 25, 26]])

In [6]:
# simillarly for 3 dimensional array
n.reshape(3,3,3)

array([[[ 0,  1,  2],
        [ 3,  4,  5],
        [ 6,  7,  8]],

       [[ 9, 10, 11],
        [12, 13, 14],
        [15, 16, 17]],

       [[18, 19, 20],
        [21, 22, 23],
        [24, 25, 26]]])

In [7]:
#converting a normal array to numpy array

m = numpy.asarray([[12, 34, 35, 23, 76],[90, 54, 16, 86, 43]])

In [8]:
type(m)

numpy.ndarray

In [10]:
m

array([[12, 34, 35, 23, 76],
       [90, 54, 16, 86, 43]])

In [12]:
import cv2       #import opencv

### Reading images with cv2
        imread method is used to read the image. 
        imread("image path", 0/1)       0 specifies gray image which is a two dimensional array
                                        1 specifies RGB image which is a three dimensional array

In [14]:
im_g = cv2.imread("smallgray.png", 0)    # read image as gray and convert to two dimensional array
im_g

array([[187, 158, 104, 121, 143],
       [198, 125, 255, 255, 147],
       [209, 134, 255,  97, 182]], dtype=uint8)

In above two dimensional array each number corresponds to the intensity of each pixel. Which ranges from 0 to 255. 
The image we have read is made up of 3 x 5 pixels (3 rows and 5 columns)

In [15]:
im_c = cv2.imread("smallgray.png", 1)    # read image as RGB and convert to three dimensional array
im_c

array([[[187, 187, 187],
        [158, 158, 158],
        [104, 104, 104],
        [121, 121, 121],
        [143, 143, 143]],

       [[198, 198, 198],
        [125, 125, 125],
        [255, 255, 255],
        [255, 255, 255],
        [147, 147, 147]],

       [[209, 209, 209],
        [134, 134, 134],
        [255, 255, 255],
        [ 97,  97,  97],
        [182, 182, 182]]], dtype=uint8)

First matrix corresponds to Blue, Second to Green and third matrix corresponds to Red. So the intesity of first pixel is (187, 198, 209) (B, G, R)

In [16]:
cv2.imwrite("newSmallGray.png", im_g)       # creates new image

True

## Indexing and slicing Numpy arrays

    array[row_lower : row_upper, column_lower: column_upper]

In [18]:
im_g

array([[187, 158, 104, 121, 143],
       [198, 125, 255, 255, 147],
       [209, 134, 255,  97, 182]], dtype=uint8)

In [26]:
im_g[0:2, 2:4]    #selected rows and columns

array([[104, 121],
       [255, 255]], dtype=uint8)

In [25]:
im_g[:, 2:4]    #all rows and selected columns

array([[104, 121],
       [255, 255],
       [255,  97]], dtype=uint8)

In [24]:
im_g[2, 4]   # one item

182

In [27]:
im_g[2:4]   #one row

array([[209, 134, 255,  97, 182]], dtype=uint8)

## Iterrating over the array

In [30]:
#iterrating rows
for i in im_g:        #iterrates over each row. It considers each row as a single item
    print(i)

[187 158 104 121 143]
[198 125 255 255 147]
[209 134 255  97 182]


In [32]:
#iterrating columns
for i in im_g.T:     #take transpose of matrix and then iterrate
    print(i)

[187 198 209]
[158 125 134]
[104 255 255]
[121 255  97]
[143 147 182]


In [33]:
#iterrating over individual item, flat method is used

for i in im_g.flat:
    print(i)

187
158
104
121
143
198
125
255
255
147
209
134
255
97
182


# Stacking and Splitting numpy array

    Stacking is the concatination of multiple arrays horizontally or vertically. For stacking the arrays the dimensions of arrays to be concatinated must be same.
    for stacking we use 'hstack' and 'vstack' methods
    
    Splitting is the dividing an array in multiple arrays with same dimensions.
    for splitting we have 'hsplit' and 'vsplit' methods.
   

In [34]:
im_g

array([[187, 158, 104, 121, 143],
       [198, 125, 255, 255, 147],
       [209, 134, 255,  97, 182]], dtype=uint8)

In [35]:
#horizontal stacking
ims = numpy.hstack((im_g, im_g))    #hstack takes only one argument. so we have to pass an tupple.
ims

array([[187, 158, 104, 121, 143, 187, 158, 104, 121, 143],
       [198, 125, 255, 255, 147, 198, 125, 255, 255, 147],
       [209, 134, 255,  97, 182, 209, 134, 255,  97, 182]], dtype=uint8)

In [38]:
#simillarly we can pass multiple arrays
ims = numpy.hstack((im_g, im_g, im_g))
print(ims)

[[187 158 104 121 143 187 158 104 121 143 187 158 104 121 143]
 [198 125 255 255 147 198 125 255 255 147 198 125 255 255 147]
 [209 134 255  97 182 209 134 255  97 182 209 134 255  97 182]]


In [39]:
#vertical stacking
imsv = numpy.vstack((im_g, im_g, im_g))
print(imsv)

[[187 158 104 121 143]
 [198 125 255 255 147]
 [209 134 255  97 182]
 [187 158 104 121 143]
 [198 125 255 255 147]
 [209 134 255  97 182]
 [187 158 104 121 143]
 [198 125 255 255 147]
 [209 134 255  97 182]]


In [41]:
#horizontal splitting

lsplt = numpy.hsplit(imsv,3)  # hsplit(array, columns in each array)

ValueError: array split does not result in an equal division

Since imsv has 5 columns it cannot be dived into 3 equal columns. hence we have to divide it in 5 equal columns

In [43]:
lsplt = numpy.hsplit(imsv, 5)
lsplt

[array([[187],
        [198],
        [209],
        [187],
        [198],
        [209],
        [187],
        [198],
        [209]], dtype=uint8), array([[158],
        [125],
        [134],
        [158],
        [125],
        [134],
        [158],
        [125],
        [134]], dtype=uint8), array([[104],
        [255],
        [255],
        [104],
        [255],
        [255],
        [104],
        [255],
        [255]], dtype=uint8), array([[121],
        [255],
        [ 97],
        [121],
        [255],
        [ 97],
        [121],
        [255],
        [ 97]], dtype=uint8), array([[143],
        [147],
        [182],
        [143],
        [147],
        [182],
        [143],
        [147],
        [182]], dtype=uint8)]

In [45]:
vsplit = numpy.vsplit(imsv, 3)
vsplit

[array([[187, 158, 104, 121, 143],
        [198, 125, 255, 255, 147],
        [209, 134, 255,  97, 182]], dtype=uint8),
 array([[187, 158, 104, 121, 143],
        [198, 125, 255, 255, 147],
        [209, 134, 255,  97, 182]], dtype=uint8),
 array([[187, 158, 104, 121, 143],
        [198, 125, 255, 255, 147],
        [209, 134, 255,  97, 182]], dtype=uint8)]

In [46]:
vsplit[0]

array([[187, 158, 104, 121, 143],
       [198, 125, 255, 255, 147],
       [209, 134, 255,  97, 182]], dtype=uint8)